DIPs/DIP1xxx-rn.md

```diff
@@ -0,0 +1,801 @@
+value would require a copy of the `struct` which would be made by calling the copy constructor, thus leading
+to an infinite chain of calls).
+
+The type qualifiers may be applied to the parameter of the copy constructor, but also to the function itself
+in order to facilitate the ability to describe copies between objects of different mutability levels. The type
+qualifiers are optional.
+
+### Semantics
+
+This section discusses all the aspects regarding the semantic analysis and interaction with other
+components of the copy constructor.
+
+#### Copy constructor and postblit cohabitation
+
+In order to assure a smooth transition from postblit to copy constructor, the following
+strategy is employed: if a `struct` defines a postblit (user-defined or generated) all
+copy constructor definitions will be ignored for that particular `struct` and the postblit
+is going to be used. Existing codebases that didn't use the postblit may start using the
+copy constructor without any problems, while codebases that did rely on the postblit may
+start writing new code using the copy constructor while removing the deprecated postblit
+from their code.
+
+Transitioning from using a postblit to using a copy constructor may simply be done by
+replacing the postblit declaration `this(this)` with `this(ref inout(typeof(this)) rhs)
+inout`.
+
+#### Copy constructor usage
+
+The copy constructor is implicitly used by the compiler whenever a struct variable is initialized
+as a copy of another variable of the same unqualified type:
+
+1. When a variable is explicitly initialized:
+
+```d
+struct A
+{
+    this(ref A rhs) {}
+}
+
+void main()
+{
+    A a;
+    A b = a; // copy constructor gets called
+    b = a;   // assignment, not initialization
+}
+```
+
+2. When a parameter is passed by value to a function:
+
+```d
+struct A
+{
+    this(ref A another) {}
+}
+
+void fun(A a) {}
+
+void main()
+{
+    A a;
+    fun(a);    // copy constructor gets called
+}
+```
+
+3. When a parameter is returned by value from a function and NRVO cannot be performed:
+
+```d
+struct A
+{
+    this(ref A another)
+    {
+        writeln("cpCtor");
+    }
+}
+
+A fun()
+{
+    A a;
+    return a;
+}
+
+A a;
+A gun()
+{
+    return a;
+}
+
+void main()
+{
+    A a;
+    a = fun();      // NRVO - no copy constructor call
+    A b;
+    b = gun();      // NRVO cannot be performed, copy constructor is called
+}
+```
+
+The parameter of the copy constructor is passed by reference, so initializations are going
+to be lowered to copy constructor calls only if the source is an lvalue. Although this can be
+worked around by declaring temporary lvalues which can be forwarded to the copy constructor,
+binding rvalues to lvalues is beyond the scope of this DIP.
+
+Note that when a function returns a struct that defines a copy constructor and NRVO cannot be
+performed, the copy constructor is called on the callee site, before actually returning. If NRVO
```

Comment:
on the callee site, before actually returning -> at the callee site before returning